### PR TITLE
Exit when last device disconnects only when when watching is disabled

### PR DESCRIPTION
**Pull request by luizoti**
Based on #63

### CODE REVIEW

**Comment by Lenbok** [comment]
I don't understand how this helps you. Your earlier comment on #63 (which now seems deleted) was about xkeysnail not detecting when you remove and re-add your "Sony PS3 BD Remote". With this PR, it just means exactly the same thing would happen if you also have another keyboard plugged in (since xkeysnail would stay running) and then plug in your Sony remote.

In other words, instead of this patch, you should fix the actual underlying problem: Why is the bluetooth keyboard not being detected while xkeysnail stays running.

**Comment by luizoti** [comment]
When I reconnect the xkeysnail is re-executed, via udev, I even prefer it, because when the control is disconnected I don't need the xkeysnail running.

As sujeri, it can be related only to bluetooth, but I don't know.

OBVIOUSLY, if xkeysnail recognizes control when reconnected, this is better behavior, but doing so is beyond my capabilities.

This correction allows the user to freely choose the behavior that best suits them.

**Comment by Lenbok** [comment]
The thing is, your patch does not appear to fix your issue in the general case, since if you have any extra keyboards (or remotes) plugged in, xkeysnail won't exit (it's only exiting if the *last* device disconnects).

I haven't seen anyone running xkeysnail via udev before - is that described somewhere? Can you share how you are doing that? Is it even being invoked with `--watch` (which is required for adding devices on re-connect)?

File: xkeysnail/input.py [hunk 1]
```diff
@@ -140,6 +140,10 @@ def loop(device_matches, device_watch, quiet):
                 if isinstance(waitable, InputDevice):
                     remove_device(devices, waitable)
                     print("Device removed: " + str(device.name))
+                    if interrupt:
+                        if not len(devices):
+                            break
+                        pass
```
Comment (Lenbok) [inline xkeysnail/input.py:146]:
```suggestion
                    if not device_watch:
                        if not len(devices):
                            break
                        pass
```

Comment (Lenbok) [reply xkeysnail/input.py:146]: If we do this, it makes more sense. Then the -i flag can be removed entirely, and exiting becomes the natural behaviour when we have not enabled watching for new device connections.

**Review by Lenbok — CHANGES_REQUESTED** [review]
My recommendation is to remove the explicit flag and exit when watching is not enabled.

File: xkeysnail/input.py (hunk 1 above)
Comment (luizoti) [reply xkeysnail/input.py:146]: I tested it and this seems to be the best way for everyone, regarding the --watch option.

I use only one control I find it efficient that the xkeysnail is closed when disconnecting and you when staying connected with more than 1 keyboard.

Well, about --watch not identifying my control, maybe this is something to be reviewed, I will test it again.

**Comment by luizoti** [comment]
> The thing is, your patch does not appear to fix your issue in the general case, since if you have any extra keyboards (or remotes) plugged in, xkeysnail won't `exit` (it's only exiting if the last device disconnects).

It also makes sense to me, but it never worked like that and since I never needed more than 1 control, in fact, even for keyboards I use, just for this control, it was never a problem.

> --watch

I tested this option, but it doesn't recognize the control as I said above, in fact, it maps my notebook keyboard, so I use a script to directly point the path /dev/input, the problem probably hangs in that area.
This operation makes sense to me, but as it occurs, it does not currently occur, looking at the script and verifying that the storage device arrives if it uses X or Y keys, my control from what I note is based on the keyboard, but there is no like keys on a keyboard.

> I haven't seen anyone running xkeysnail via udev before

I use use udev rules.

**Review by Lenbok — APPROVED** [review]
This looks good to me now.

Just to tidy up, I would suggest updating the PR title to something like "Exit when last device disconnects only when when watching is disabled"

**Pull request closed without merging**